src/balloonpanel/balloonpanelview.js

```diff
@@ -177,30 +178,37 @@ export default class BalloonPanelView extends View {
 			} )
 		};
 
-		this._smartAttachTo( possiblePanelRects, visibleRect );
+		this._smartAttachTo( possiblePanelRects, limiterVisibleRect, panelRect.width * panelRect.height );
 	}
 
 	/**
-	 * For the given `Array` of possible rects, choses the one which fits the best into
-	 * `visibleViewportRect`, which is when their intersection has the biggest area.
+	 * For the given set of possible rects, chooses the one which fits the best into both - browser viewport and
+	 * `visibleContainerRect`, which is when their intersection has the biggest area. Note that priority is a possible
+	 * highest intersection area with browser viewport.
 	 *
 	 * @private
-	 * @param {Array<{AbsoluteDomRect}>} rects List of positions where balloon can be placed.
-	 * @param {AbsoluteDomRect} visibleViewportRect The absolute rect of the visible part of the browser viewport.
+	 * @param {Object} rects Set of positions where balloon can be placed.
+	 * @param {AbsoluteDomRect} visibleContainerRect The absolute rect of the visible part of container element.
 	 */
-	_smartAttachTo( rects, visibleViewportRect ) {
+	_smartAttachTo( rects, visibleContainerRect, panelSurfaceArea ) {
+		const viewportRect = new AbsoluteDomRect( getViewportRect() );
 		let maxIntersectRectPos;
-		let maxIntersectArea = -1;
+		let maxContainerIntersectArea = -1;
+		let maxViewportIntersectArea = -1;
 
 		// Get best place.
-		for ( let rectPos in rects ) {
-			const intersectArea = rects[ rectPos ].getIntersectArea( visibleViewportRect );
+		Object.keys( rects ).some( ( rectPos ) => {
+			const containerIntersectArea = rects[ rectPos ].getIntersectArea( visibleContainerRect );
+			const viewportIntersectArea = rects[ rectPos ].getIntersectArea( viewportRect );
 
-			if ( intersectArea > maxIntersectArea ) {
+			if ( viewportIntersectArea >= maxViewportIntersectArea && containerIntersectArea > maxContainerIntersectArea ) {
 				maxIntersectRectPos = rectPos;
-				maxIntersectArea = intersectArea;
+				maxContainerIntersectArea = containerIntersectArea;
+				maxViewportIntersectArea = viewportIntersectArea;
 			}
-		}
+
+			return maxContainerIntersectArea === panelSurfaceArea;
```

Comment:
1. I was thinking about it not in case of performance, but in case of position priorities. You are right, it does not improve the performance that much.

2. `maxContainerIntersectArea` is equal to `panelSurfaceArea` only when is fully visible in viewport
```javascript
const containerIntersectArea = rects[ rectPos ].getIntersectArea( visibleContainerRect ); // visibleContainerRect
```